src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -5495,6 +5527,332 @@ register_handler(int sig, struct sigaction *old_handler)
+	if (io_depth == 0)
+		return;
+
+	pthread_mutex_lock(&lock_aio);
+	if (aio_ctx_list[ctx_idx].inited) {
+		/* if eq initialized */
+		for (i = 0; i < aio_ctx_list[ctx_idx].depth; i++) {
+			daos_eq_destroy(aio_ctx_list[ctx_idx].eq_ev_list[i].eq, 0);
+		}
+		D_FREE(aio_ctx_list[ctx_idx].eq_ev_list);
+		aio_eq_count -= aio_ctx_list[ctx_idx].depth;
+	}
+	aio_ctx_list[ctx_idx].ctx = 0;
+	aio_ctx_count--;
+	pthread_mutex_unlock(&lock_aio);
+}
+
+int
+io_setup(int maxevents, io_context_t *ctxp)
+{
+	int rc;
+
+	if (next_io_setup == NULL) {
+		next_io_setup = dlsym(RTLD_NEXT, "io_setup");
+		D_ASSERT(next_io_setup != NULL);
+	}
+
+	rc = next_io_setup(maxevents, ctxp);
+	if (rc < 0)
+		return rc;
+
+	allocate_aio_ctx(*ctxp, maxevents);
+	/* we ignore the result value here since we are not sure whether DFS files will
+	 * be involved or not.
+	 */
+	return rc;
+}
+
+int
+io_destroy(io_context_t ctx)
+{
+	if (next_io_destroy == NULL) {
+		next_io_destroy = dlsym(RTLD_NEXT, "io_destroy");
+		D_ASSERT(next_io_destroy != NULL);
+	}
+	free_aio_ctx(ctx);
+	return next_io_destroy(ctx);
+}
+
+static int
+create_eq_ev_for_aio(int ctx_idx, int io_depth)
+{
+	int i, j, rc;
+
+	if (aio_ctx_list[ctx_idx].inited) {
+		if (io_depth <= aio_ctx_list[ctx_idx].depth) {
+			return 0;
+		} else {
+			DS_ERROR(EINVAL, "the number of requested ops is larger then the context "
+					 "depth");
+			return EINVAL;
+		}
+	}
+
+	if ((eq_count_max - eq_count - aio_eq_count) < io_depth) {
+		DS_ERROR(EBUSY, "no enough eq to create requested contexts");
+		return EBUSY;
+	}
+
+	/* to create more eq */
+	pthread_mutex_lock(&lock_eqh);
+	D_ALLOC_ARRAY(aio_ctx_list[ctx_idx].eq_ev_list, io_depth);
+	if (aio_ctx_list[ctx_idx].eq_ev_list == NULL) {
+		pthread_mutex_unlock(&lock_eqh);
+		return ENOMEM;
+	}
+
+	for (i = 0; i < io_depth; i++) {
+		aio_ctx_list[ctx_idx].eq_ev_list[i].queued_op_count = 0;
+		rc = daos_eq_create(&aio_ctx_list[ctx_idx].eq_ev_list[i].eq);
+		if (rc)
+			goto free_eq;
+	}
+
+	aio_eq_count += io_depth;
+	aio_ctx_list[ctx_idx].inited = true;
+	pthread_mutex_unlock(&lock_eqh);
+
+	return 0;
+
+free_eq:
+	for (j = 0; j < i; j++) {
+		daos_eq_destroy(aio_ctx_list[ctx_idx].eq_ev_list[j].eq, 0);
+	}
+	D_FREE(aio_ctx_list[ctx_idx].eq_ev_list);
+	pthread_mutex_unlock(&lock_eqh);
+	return daos_der2errno(rc);
+}
+
+int
+io_submit(io_context_t ctx, long nr, struct iocb *ios[])
+{
+	int         i, n_op_dfs, fd, io_depth, ctx_idx;
+	daos_size_t read_size = 0;
+	d_iov_t     iov       = {};
+	d_sg_list_t sgl       = {};
+	int         rc;
+	short       op;
+
+	if (next_io_submit == NULL) {
+		next_io_submit = dlsym(RTLD_NEXT, "io_submit");
+		D_ASSERT(next_io_submit != NULL);
+	}
+	if (!hook_enabled)
+		return next_io_submit(ctx, nr, ios);
+	io_depth = query_aio_ctx_depth(ctx, &ctx_idx);
+	if (io_depth == 0)
+		return next_io_submit(ctx, nr, ios);
+	if (nr > io_depth) {
+		DS_ERROR(EINVAL, "number of ops submitted is larger than queue size");
```

Comment:
is this correct? EINVAL?
don't we submit the ones we can and return that?

Reply:
You are right. I will fix it. Thank you!